addons/website/tests/test_lang_url.py

```diff
@@ -28,30 +30,40 @@ def test_02_url_redirect(self):
         url = '/fr_WHATEVER/contactus'
         r = self.url_open(url)
         self.assertEqual(r.status_code, 200)
-        self.assertTrue(r.url.endswith('/fr/contactus'), "fr_WHATEVER should be forwarded to 'fr_FR' lang as closest match")
+        self.assertTrue(r.url.endswith('/fr/contactus'), f"fr_WHATEVER should be forwarded to 'fr_FR' lang as closest match, url: {r.url}")
 
         url = '/fr_FR/contactus'
         r = self.url_open(url)
         self.assertEqual(r.status_code, 200)
-        self.assertTrue(r.url.endswith('/fr/contactus'), "lang in url should use url_code ('fr' in this case)")
+        self.assertTrue(r.url.endswith('/fr/contactus'), f"lang in url should use url_code ('fr' in this case), url: {r.url}")
 
     def test_03_url_cook_lang_not_available(self):
         """ An activated res.lang should not be displayed in the frontend if not a website lang. """
         self.website.language_ids = self.env.ref('base.lang_en')
         r = self.url_open('/fr/contactus')
-        self.assertTrue('lang="en-US"' in r.text, "french should not be displayed as not a frontend lang")
+        doc = lxml.html.document_fromstring(r.text)
+        self.assertEqual(doc.get('lang'), 'en-US', "french should not be displayed as not a frontend lang")
 
     def test_04_url_cook_lang_not_available(self):
         """ `nearest_lang` should filter out lang not available in frontend.
-        Eg: 1. go in backend in english -> request.context['lang'] = `en_US`
-            2. go in frontend, the request.context['lang'] is passed through
+        Eg: 1. go in backend in english -> request.env.context['lang'] = `en_US`
+            2. go in frontend, the request.env.context['lang'] is passed through
                `nearest_lang` which should not return english. More then a
                misbehavior it will crash in website language selector template.
         """
         # 1. Load backend
         self.authenticate('admin', 'admin')
         r = self.url_open('/web')
-        self.assertTrue('"lang": "en_US"' in r.text, "ensure english was loaded")
+        self.assertEqual(r.status_code, 200)
+
+        for line in r.text.splitlines():
+            _, match, session_info_str = line.partition('odoo.__session_info__ = ')
+            if match:
+                session_info = json.loads(session_info_str[:-1])
+                self.assertEqual(session_info['user_context']['lang'], 'en_US', "ensure english was loaded")
+                break
+        else:
+            raise ValueError('Session info not found in web page')
```

Comment:
Same as above, it is used to produce a better fail report because I wasn't  productive with "False is not True". `str.splitlines()` + `str.partition()` + 1x `json.loads()` is not too much slower from `str.__contains__()`, I expect a few milli-seconds diff which I find acceptable to produce a better error message.

Reply:
but using old code, you know what was tested :fearful: